templates/bucket-replication-lambda.yaml

```diff
@@ -0,0 +1,200 @@
+      Code:
+        S3Bucket: !Ref FunctionCodeBucket
+        S3Key: !Ref FunctionCodeKey
+      Description: Adds files from an S3 bucket to Synapse.
+      Environment:
+        Variables:
+          SYNAPSE_USERNAME: !Ref SynapseUsername
+          SYNAPSE_APIKEY: !Ref SynapseApikey
+          SYNAPSE_SOURCE_BUCKET_STORAGE_LOCATION_ID: !Ref SynapseStorageLocationId
+          SYNAPSE_NOTIFY_USERS: !Ref SynapseNotifyUsers
+          MD5_BLOCK_SIZE: !Ref Md5BlockSize
+      Handler: lambda_function.lambda_handler
+      Runtime: python3.7
+      Timeout: 300
+      Role: !GetAtt LambdaExecutionRole.Arn
+
+  DataBucket:
+    Type: AWS::S3::Bucket
+    DeletionPolicy: Delete
+    DependsOn:
+      - LambdaExecutionRolePolicy
+    Properties:
+      BucketName: !Sub ${AWS::StackName}-data
+      VersioningConfiguration:
+        Status: Enabled
+      NotificationConfiguration:
+        LambdaConfigurations:
+          - Event: s3:ObjectCreated:*
+            Function: !GetAtt [BucketToSynapseFunction, Arn]
+
+  BucketPermission:
+    Type: AWS::Lambda::Permission
+    Properties:
+      Action: 'lambda:InvokeFunction'
+      FunctionName: !Ref BucketToSynapseFunction
+      Principal: s3.amazonaws.com
+      SourceAccount: !Ref 'AWS::AccountId'
+      SourceArn: !Join
+          - ""
+          - - "arn:aws:s3:::"
+            - !Sub ${AWS::StackName}-data
+
+  BucketPolicy:
+    Type: AWS::S3::BucketPolicy
+    Properties:
+      Bucket: !Ref DataBucket
+      PolicyDocument:
+        Version: '2012-10-17'
+        Id: PolicyForDestinationBucket
+        Statement:
+          # Allow Synapse to access the data bucket
+          - Sid: Synapse1
+            Principal:
+              AWS: !Sub "arn:aws:iam::${SynapseAWSAccountId}:root"
+            Action:
+              - "s3:ListBucket*"
+              - "s3:GetBucketLocation"
+            Resource: !GetAtt
+              - DataBucket
+              - Arn
+            Effect: 'Allow'
+          - Sid: Synapse2
+            Principal:
+              AWS: !Sub "arn:aws:iam::${SynapseAWSAccountId}:root"
+            Action:
+              - "s3:GetObject*"
+              - "s3:*MultipartUpload*"
+            Resource: !Join [ "", [!GetAtt [DataBucket, Arn], "/*"]]
+            Effect: "Allow"
+          # Allow the source bucket to write to the data bucket
+          - Sid: SourceAccount1
+            Effect: Allow
+            Principal:
+              AWS: !Sub arn:aws:iam::${ReplicationSourceAccountId}:root
+            Action:
+              - s3:GetBucketVersioning
+              - s3:ReplicateObject
+              - s3:ReplicateDelete
+              - s3:ObjectOwnerOverrideToBucketOwner
+            Resource:
+              - !GetAtt [DataBucket, Arn]
+              - !Join [ "", [!GetAtt [DataBucket, Arn], "/*"]]
+
+
+Outputs:
+  BucketToSynapseFunctionArn:
+    Description: ARN of the lambda function that was created in this template.
+    Value: !GetAtt BucketToSynapseFunction.Arn
```

Comment:
we usually like to add a Export to each Output so we can reference the values across templates.  i.e. 
```
    Export:
      Name: !Sub '${AWS::Region}-${AWS::StackName}-BucketToSynapseFunctionArn'
```